File: packages/evaluatorq/src/lib/tracing/spans.ts
/**
 * Span creation utilities for OpenTelemetry instrumentation.
 *
 * Span hierarchy:
 * - orq.evaluation_run (root or child of parent context)
 *   └── orq.job (per job per data point)
 *       ├── [User's instrumented code becomes child spans]
 *       └── orq.evaluation (per evaluator - child of its job)
 */

import type {
  Context,
  Span,
  SpanStatusCode as SpanStatusCodeType,
} from "@opentelemetry/api";

import { getTracer } from "./setup.js";

/**
 * Execute a callback within a span, handling status and error recording.
 */
async function executeWithSpan<T>(
  span: Span,
  SpanStatusCode: typeof SpanStatusCodeType,
  fn: (span: Span) => Promise<T>,
): Promise<T> {
  try {
    const result = await fn(span);
    span.setStatus({ code: SpanStatusCode.OK });
    return result;
  } catch (error) {
    span.setStatus({
      code: SpanStatusCode.ERROR,
      message: error instanceof Error ? error.message : String(error),
    });
    span.recordException(
      error instanceof Error ? error : new Error(String(error)),
    );
    throw error;
  } finally {
    span.end();
  }
}

export interface EvaluationRunSpanOptions {
  runId: string;
  runName: string;
  dataPointsCount: number;
  jobsCount: number;
  evaluatorsCount: number;
  parentContext?: Context;
}

export interface JobSpanOptions {
  runId: string;
  rowIndex: number;
  jobName?: string;
  parentContext?: unknown;
}

export interface EvaluationSpanOptions {
  runId: string;
  evaluatorName: string;
}

/**
 * Execute a function within an orq.evaluation_run span.
 * This is the root span for an evaluation run.
 */
export async function withEvaluationRunSpan<T>(
  options: EvaluationRunSpanOptions,
  fn: (span: Span | undefined) => Promise<T>,
): Promise<T> {
  const tracer = getTracer();
  if (!tracer) {
    return fn(undefined);
  }

  try {
    const { context, SpanStatusCode } = await import("@opentelemetry/api");

    // Use parent context if provided, otherwise use active context
    const parentCtx = options.parentContext || context.active();

    return await tracer.startActiveSpan(
      "orq.evaluation_run",
      {
        attributes: {
          "orq.run_id": options.runId,
          "orq.run_name": options.runName,
          "orq.data_points_count": options.dataPointsCount,
          "orq.jobs_count": options.jobsCount,
          "orq.evaluators_count": options.evaluatorsCount,
        },
      },
      parentCtx,
      (span) => executeWithSpan(span, SpanStatusCode, fn),
    );
  } catch {
    // OTEL not available, run without span
    return fn(undefined);
  }
}

/**
 * Execute a function within an orq.job span.
 * Job spans are independent roots, or children of a parent context if provided.
 */
export async function withJobSpan<T>(
  options: JobSpanOptions,
  fn: (span: Span | undefined) => Promise<T>,
): Promise<T> {
  const tracer = getTracer();
  if (!tracer) {
    return fn(undefined);
  }

  try {
    const { context, SpanStatusCode } = await import("@opentelemetry/api");

    // Use parent context if provided, otherwise use active context
    const parentCtx = (options.parentContext as Context) || context.active();

    return await tracer.startActiveSpan(
      "orq.job",
      {
        attributes: {
          "orq.run_id": options.runId,
          "orq.row_index": options.rowIndex,
          ...(options.jobName && { "orq.job_name": options.jobName }),
        },
      },
      parentCtx,
      (span) => executeWithSpan(span, SpanStatusCode, fn),
    );
  } catch {
    // OTEL not available, run without span
    return fn(undefined);
  }
}

/**
 * Execute a function within an orq.evaluation span.
 * Evaluation spans are children of the job span.
 */
export async function withEvaluationSpan<T>(
  options: EvaluationSpanOptions,
  fn: (span: Span | undefined) => Promise<T>,
): Promise<T> {
  const tracer = getTracer();
  if (!tracer) {
    return fn(undefined);
  }

  try {
    const { SpanStatusCode } = await import("@opentelemetry/api");

    return await tracer.startActiveSpan(
      "orq.evaluation",
      {
        attributes: {
          "orq.run_id": options.runId,
          "orq.evaluator_name": options.evaluatorName,
        },
      },
      (span) => executeWithSpan(span, SpanStatusCode, fn),
    );
  } catch {
    // OTEL not available, run without span
    return fn(undefined);
  }
}

/**
 * Set evaluation result attributes on a span.
 */
export function setEvaluationAttributes(
  span: Span | undefined,
  score: string | number | boolean | Record<string, unknown>,
  explanation?: string,
  pass?: boolean,
): void {
  if (!span) return;

  span.setAttribute(
    "orq.score",
    typeof score === "object" && score !== null
      ? JSON.stringify(score)
      : String(score),
  );
  if (explanation !== undefined) {
    span.setAttribute("orq.explanation", explanation);
  }
  if (pass !== undefined) {
    span.setAttribute("orq.pass", pass);
  }
}

/**
 * Set the job name attribute on a span after job execution.
 */
export function setJobNameAttribute(
  span: Span | undefined,
  jobName: string,
): void {
  if (!span) return;
  span.setAttribute("orq.job_name", jobName);
}

File: packages/evaluatorq-py/src/evaluatorq/tracing/spans.py
"""
Span creation utilities for OpenTelemetry instrumentation.

Span hierarchy:
- orq.evaluation_run (root or child of parent context)
  └── orq.job (per job per data point)
      ├── [User's instrumented code becomes child spans]
      └── orq.evaluation (per evaluator - child of its job)
"""
from __future__ import annotations

import json

from collections.abc import AsyncGenerator
from contextlib import asynccontextmanager
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from .setup import get_tracer
from ..types import EvaluationResultCell

if TYPE_CHECKING:
    from opentelemetry.trace import Span


@dataclass
class EvaluationRunSpanOptions:
    """Options for creating an evaluation run span."""

    run_id: str
    run_name: str
    data_points_count: int
    jobs_count: int
    evaluators_count: int
    parent_context: Any | None = None


@dataclass
class JobSpanOptions:
    """Options for creating a job span."""

    run_id: str
    row_index: int
    job_name: str | None = None
    parent_context: Any | None = None


@dataclass
class EvaluationSpanOptions:
    """Options for creating an evaluation span."""

    run_id: str
    evaluator_name: str


@asynccontextmanager
async def with_evaluation_run_span(
    options: EvaluationRunSpanOptions,
) -> AsyncGenerator["Span | None", None]:
    """
    Execute code within an orq.evaluation_run span.
    This is the root span for an evaluation run.

    Args:
        options: Evaluation run span configuration

    Yields:
        The span if tracing is enabled, None otherwise
    """
    tracer = get_tracer()
    if tracer is None:
        yield None
        return

    # Import OTEL dependencies first, before entering the span.
    # This keeps the ImportError handler narrow — it only catches
    # missing OTEL packages, never errors from user code inside the span.
    try:
        from opentelemetry import context as otel_context
        from opentelemetry.trace import SpanKind, Status, StatusCode
    except ImportError:
        yield None
        return

    parent_ctx = options.parent_context or otel_context.get_current()

    with tracer.start_as_current_span(
        "orq.evaluation_run",
        context=parent_ctx,
        kind=SpanKind.INTERNAL,
        attributes={
            "orq.run_id": options.run_id,
            "orq.run_name": options.run_name,
            "orq.data_points_count": options.data_points_count,
            "orq.jobs_count": options.jobs_count,
            "orq.evaluators_count": options.evaluators_count,
        },
    ) as span:
        try:
            yield span
            span.set_status(Status(StatusCode.OK))
        except Exception as e:
            span.set_status(Status(StatusCode.ERROR, str(e)))
            span.record_exception(e)
            raise


@asynccontextmanager
async def with_job_span(
    options: JobSpanOptions,
) -> AsyncGenerator["Span | None", None]:
    """
    Execute code within an orq.job span.
    Job spans are independent roots, or children of a parent context if provided.

    Args:
        options: Job span configuration

    Yields:
        The span if tracing is enabled, None otherwise

    Example:
        async with with_job_span(JobSpanOptions(run_id="abc", row_index=0)) as span:
            # Your job code here
            pass
    """
    tracer = get_tracer()
    if tracer is None:
        yield None
        return

    try:
        from opentelemetry import context as otel_context
        from opentelemetry.trace import SpanKind, Status, StatusCode

        # Use parent context if provided, otherwise use active context
        parent_ctx = options.parent_context or otel_context.get_current()

        attributes: dict[str, Any] = {
            "orq.run_id": options.run_id,
            "orq.row_index": options.row_index,
        }
        if options.job_name:
            attributes["orq.job_name"] = options.job_name

        with tracer.start_as_current_span(
            "orq.job",
            context=parent_ctx,
            kind=SpanKind.INTERNAL,
            attributes=attributes,
        ) as span:
            try:
                yield span
                span.set_status(Status(StatusCode.OK))
            except Exception as e:
                span.set_status(Status(StatusCode.ERROR, str(e)))
                span.record_exception(e)
                raise

    except ImportError:
        # OTEL not available, run without span
        yield None


@asynccontextmanager
async def with_evaluation_span(
    options: EvaluationSpanOptions,
) -> AsyncGenerator["Span | None", None]:
    """
    Execute code within an orq.evaluation span.
    Evaluation spans are children of the job span.

    Args:
        options: Evaluation span configuration

    Yields:
        The span if tracing is enabled, None otherwise

    Example:
        async with with_evaluation_span(EvaluationSpanOptions(
            run_id="abc",
            evaluator_name="string-contains"
        )) as span:
            # Your evaluator code here
            pass
    """
    tracer = get_tracer()
    if tracer is None:
        yield None
        return

    try:
        from opentelemetry.trace import SpanKind, Status, StatusCode

        with tracer.start_as_current_span(
            "orq.evaluation",
            kind=SpanKind.INTERNAL,
            attributes={
                "orq.run_id": options.run_id,
                "orq.evaluator_name": options.evaluator_name,
            },
        ) as span:
            try:
                yield span
                span.set_status(Status(StatusCode.OK))
            except Exception as e:
                span.set_status(Status(StatusCode.ERROR, str(e)))
                span.record_exception(e)
                raise

    except ImportError:
        # OTEL not available, run without span
        yield None


def set_evaluation_attributes(
    span: "Span | None",
    score: str | int | float | bool | dict[str, Any] | EvaluationResultCell,
    explanation: str | None = None,
    pass_: bool | None = None,
) -> None:
    """
    Set evaluation result attributes on a span.

    Args:
        span: The span to set attributes on (can be None)
        score: The evaluation score
        explanation: Optional explanation of the score
        pass_: Optional pass/fail status
    """
    if span is None:
        return

    span.set_attribute(
        "orq.score",
        json.dumps(score.model_dump()) if isinstance(score, EvaluationResultCell) else json.dumps(score) if isinstance(score, dict) else str(score),
    )
    if explanation is not None:
        span.set_attribute("orq.explanation", explanation)
    if pass_ is not None:
        span.set_attribute("orq.pass", pass_)


def set_job_name_attribute(span: "Span | None", job_name: str) -> None:
    """
    Set the job name attribute on a span after job execution.

    Args:
        span: The span to set the attribute on (can be None)
        job_name: The name of the job
    """
    if span is None:
        return
    span.set_attribute("orq.job_name", job_name)
